### PR TITLE
pause / resume support

### PR DESCRIPTION
**Pull request by mfussenegger**

### CODE REVIEW

File: sql/src/main/java/io/crate/operation/MultiUpstreamRowDownstream.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import io.crate.core.collections.Row;
+import org.elasticsearch.common.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class MultiUpstreamRowDownstream implements RowDownstream, RowDownstreamHandle {
+
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/MultiUpstreamRowDownstream.java]: does this need to implement rowdownstreamhandle?
why not store the state in handles?

File: sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java [hunk 2]
```diff
@@ -161,29 +158,59 @@ public void collect(int doc) throws IOException {
...
             throw new CollectionFinishedEarlyException();
         }
     }
 
+    protected boolean skipDoc(int doc) {
+        ScoreDoc after = internalCollectContext.lastCollected;
+        if (after != null) {
+            if (currentDocBase == internalCollectContext.lastDocBase && doc <= after.doc) {
+                // doc was already collected, skip
+                LOGGER.trace("skipping doc {} <= {} of docBase {}", doc, after.doc, currentDocBase);
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java]: do we really want to log here?

File: sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java [hunk 3]
```diff
@@ -161,29 +158,59 @@ public void collect(int doc) throws IOException {
...
         }
     }
 
-    public void setNextOrderByValues(ScoreDoc scoreDoc) {
-        for (OrderByCollectorExpression expr : orderByCollectorExpressions) {
-            expr.setNextFieldDoc((FieldDoc) scoreDoc);
+    protected void skipSegmentReader() {
+        if (currentDocBase < internalCollectContext.lastDocBase) {
+            // skip this segment reader, all docs of this segment are already collected
+            LOGGER.trace("skipping segment reader with docBase {} < {}",
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java]: also here no logging pls

File: sql/src/main/java/io/crate/operation/collect/OrderedLuceneDocCollector.java [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+            if (orderBy == null) {
+                return null;
+            }
+            BooleanQuery query = new BooleanQuery();
+            for (int i = 0; i < orderBy.orderBySymbols().size(); i++) {
+                Symbol order = orderBy.orderBySymbols().get(i);
+                Object value = lastCollected.fields[i];
+                // only filter for null values if nulls last
+                if (order instanceof Reference && (value != null || !orderBy.nullsFirst()[i])) {
+                    QueryBuilderHelper helper = QueryBuilderHelper.forType(order.valueType());
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/collect/OrderedLuceneDocCollector.java]: would probably make sense to do something of this builder stuff outside the page loop

**Comment by mfussenegger** [comment]
Merged https://github.com/crate/crate/pull/2452 instead

**Pull request closed without merging**